Add example DAG for demonstrating usage of GCS sensors

pankajkoti opened the pull request:
Following GCS Sensor examples are provided as part of the change:
1. GCSUploadSessionCompleteSensor
2. GCSObjectUpdateSensor

Commit:
Add example DAG for demonstrating usage of GCS sensors

Following GCS Sensors examples are provided as part of the change:
1. GCSUploadSessionCompleteSensor
2. GCSObjectUpdateSensor

phanikumv reviewed (commented):
Looks good to me

kaxil commented on airflow/example_dags/example_gcs.py:
```diff
@@ -0,0 +1,74 @@
+    catchup=False,
+    default_args=DEFAULT_ARGS,
+    tags=["example", "gcs"],
+) as dag:
```
```suggestion
with DAG(
    "example_gcs_sensors",
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    schedule_interval=None,
    catchup=False,
    default_args={
        "execution_timeout": timedelta(minutes=30),
    },
    tags=["example", "gcs"],
) as dag:
```

kaxil replied:
Since we only use it once

kaxil commented on airflow/example_dags/example_gcs.py:
```diff
@@ -0,0 +1,74 @@
+    tags=["example", "gcs"],
+) as dag:
+
+    # [START howto_sensor_gcs_upload_session_complete_task]
```
Should we add this in https://airflow.apache.org/docs/apache-airflow-providers-google/stable/operators/cloud/gcs.html ?

You can add it in https://github.com/apache/airflow/blob/main/docs/apache-airflow-providers-google/operators/cloud/gcs.rst which will show up in the page I linked above when new version of provider is released.

You can also build docs locally as described in https://github.com/apache/airflow/blob/main/BREEZE.rst#building-the-documentation

```
./breeze build-docs -- --package-filter apache-airflow-providers-google
```

pankajkoti committed:
Update airflow/example_dags/example_gcs.py

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>

pankajkoti committed:
Merge branch 'main' into examples-gcs-dag

pankajkoti committed:
Delete new DAG file and add sensors to existing example_gcs.py

The commit does the following:
1. Delete the newly created top level example_gcs.py as it was a
   wrong place for the sensors
2. Add the intended sensors of the PR to the existing example_gcs.py file
   located in airflow/cloud/example_dags directory

pankajkoti committed:
Update documentation for new sensor examples

pankajkoti committed:
Update docstring to resolve spell check error in docs generation

pankajkoti committed:
Merge branch 'main' into examples-gcs-dag

pankajkoti committed:
Merge branch 'main' into examples-gcs-dag

pankajkoti committed:
Reorder imports to fix failing Isort static check

pankajkoti committed:
Merge branch 'examples-gcs-dag' of https://github.com/astronomer/airflow into examples-gcs-dag

bhirsz commented on airflow/providers/google/cloud/example_dags/example_gcs.py:
```diff
@@ -37,35 +37,44 @@
 
 BUCKET_FILE_LOCATION = PATH_TO_UPLOAD_FILE.rpartition("/")[-1]
 
+# Upload 'test-gcs-manual-example-upload.txt' manually in the <BUCKET_1> after triggering the DAG.
```
Tests shouldn't depend on manual steps - if the file is required we should:

1) store it in resources
2) upload using operator, for example LocalFilesystemToGCSOperator
3) After the tests, remove any resources created during tests (usually it's enough to remove Bucket)

pankajkoti replied:
Hi @bhirsz I wanted to do upload the file programatically like the other chain. However, the sensor task GCSUploadSessionCompleteSensor waits for change in number of files in the bucket. If we have the upload task before GCSUploadSessionCompleteSensor task, it won't detect any changes. On the other hand, if we add upload task after GCSUploadSessionCompleteSensor task, it would be blocked until GCSUploadSessionCompleteSensor task completes and does not solve the need. I am unsure how to add dependency between such tasks. Any suggestions?

bhirsz replied:
It's tricky in this case, indeed. How about starting sensors in parallel with upload_file? 

Example:
```
    chain(
        # TEST SETUP
        create_bucket,
        upload_file,
        # TEST BODY
        [gcs_object_exists, gcs_object_with_prefix_exists],
        # TEST TEARDOWN
        delete_bucket,
    )
    chain(
        create_bucket,
        # TEST BODY
        [gcs_upload_session_complete, gcs_update_object_exists],
        delete_bucket
    )
```

![image](https://user-images.githubusercontent.com/8532066/162905393-61247c98-9661-40b3-97db-9a2f2aeb660c.png)

We're starting the sensors and in meantime we're uploading the file - and sensors detect it:
![image](https://user-images.githubusercontent.com/8532066/162905681-eab7e6a1-967b-43ac-9582-aac5e21dea16.png)

![image](https://user-images.githubusercontent.com/8532066/162905856-f8a093b6-85cf-46d3-b2f6-461e2dbbe51a.png)

I'm not sure though if it will not be flaky in some cases - running this in CI will show. I will update my PR with this change.

pankajkoti replied:
Right. Since, we are starting in parallel, it may happen that, the upload task is picked up before the sensor task begins and it may not detect the change as expected. Hence, added the comment to manually upload the file. :)

Also, the gcs_object_update_sensor_task needs to be activated after gcs_upload_session_complete_task (and not in parallel with it) as the object is expected to be detected by sensor and the object_update task is to confirm the manually upload has happened prior to it.

bhirsz replied:
We can alleviate it a bit with sleep (yes, the old the ugly sleep ;)). Run sensors separably and in the meantime trigger sleep (5s will suffice) and then upload task:


    chain(
        # TEST SETUP
        create_bucket,
        sleep,
        upload_file,
        # TEST BODY
        [gcs_object_exists, gcs_object_with_prefix_exists],
        # TEST TEARDOWN
        delete_bucket,
    )
    chain(
        create_bucket,
        # TEST BODY
        gcs_upload_session_complete,
        gcs_update_object_exists,
        delete_bucket
    )

And of course put some explanation in comments why we're doing it.

bhirsz replied:
Manual steps are unfortunately a big no for us since the plan is to run system tests in CI in community. We can't expect people performing manual task while running them so I'm looking for an automatic solution

pankajkoti replied:
Okay, I was not aware of the fact that we're migrating these example DAGs to tests. I was of the impression that the example DAGs are for references to our community on how to implement DAGs for certain operators. Understand and totally agree to your fair point that tests should not need manual interventions. 
Okay, so we can try introducing sleep if we feel right about it. Also, I am new to contributing to Airflow :) 

cc: @kaxil

bhirsz replied:
Yeah, we had two goals with the migration - to make writing system tests easier and actually ensure that are example dags are runnable (what would be point in example that doesn't work, and it's actually often the case ;)). In the old design system tests actually "wrapped" and run examples - but not all example dags were used by system tests. In the new design we're also using examples as system tests, but without separating it to different files and now example dag is the system test itself. 

Any change is highly welcome and it's good to see PRs such like yours - congrats! We only missed the notification on PR (sadly Github doesn't allow for an advanced notification system) and we were not able to have this discussion before merging the PR.
